가끔은 setUp()에서 외부 자원을 할당하는 경우가 있다.
테스트가 계속 서로 독립적이길 바란다면 외부 자원을 할당받은 테스트들은 작업을 마치기 전에 tearDown() 메서드 같은 곳에서 자원을 다시 반환할 필요가 있다.

단순하게 생각하면 할당 해제를 위한 테스트 방법은 역시 또 하나의 플래그를 도입하는 것이다.
하지만 이제는 플래그들이 귀찮게 만들기 시작한다.
그리고 플래그를 사용하는 방식은 메서드의 중요한 면을 하나 놓치고 있다.
바로 setUp()은 테스트 메서드가 실행되기 전에 호출되어야 하고 tearDown()은 테스트 메서드가 실행된 후에 호출되어야 한다는 점이다.

난 호출된 메서드의 로그를 간단히 남기는 식으로 테스트 전략을 바꿀 생각이다.
항상 로그의 끝부분에만 기록을 추가하면 메서드 호출 순서를 알 수 있게 될 것이다.
```python
# WasRun
def setUp(self):
    self.wasRun = None
    self.wasSetUp = 1
    self.log = "setUp"
```
이제 testSetUp()이 플래그 대신에 로그를 검사하도록 변경할 수 있다.
```python
# TestCaseTest
def testSetUp(self):
    self.test.run()
    assert "setUp " == self.test.log
```
다음엔 wasSetup 플래그를 지울 수 있다.
```python
# WasRun
def testMethod(self):
    self.wasRun = 1
    self.log = self.log + "testMethod "
```
이 작업은 testSetUp을 실패하게 만든다.
왜냐하면 실제 로그는 'setUpTestMethod'이기 때문이다. 예상되는 값 부분을 바꾸자.
```python
# TestCaseTest
def testSetUp(self):
    self.test.run()
    assert "setUp testMethod " == self.test.log
```
이제 이 테스트는 두 개의 테스트가 할 일을 모두 수행한다. 따라서 testRunning을 지ㅜㅇ고 testSetUp의 이름을 바꿔주자.
```python
# TestCaseTest
def setUp(self):
    self.test = WasRun("testMethod")
def testTemplateMethod(self):
    self.test.run()
    assert("setUp testMethod " == self.test.log)
```

tearDown() 테스트
```python
# TestCaseTest
def testTemplateMethod(self):
    test = WasRun("testMethod")
    test.run()
    assert("setUp testMethod tearDown " == self.test.log)
```
실패한다. 성공하게 만드는 것은 간단하다.
```python
# TestCase
def run(self):
    self.setUp()
    method = getattr(self, self.name)
    method()
    self.tearDown()
```
```python
# WasRun
def setUp(self):
    self.log = "setUp "
def testMethod(self):
    self.log = self.log + "testMethod "
def tearDown(self):
    self.log = self.log + "tearDown"
```
```python
# TestCase
def tearDown(self):
    pass
```
이번에는 우리가 지금 개발하는 동일한 테스팅 프레임워크를 써먹어서 이득이 있었다.

**검토**
- 플래그에서 로그로 테스트 전략을 구조 조정했다.
- 새로운 로그 기능을 이용하여 tearDown()을 테스트하고 구현했다.
- 문제를 발견했는데 뒤로 되돌아가는 대신 과감히 수정했다. (이게 잘 한 일일까?)